docs/usage.rst

```diff
@@ -0,0 +1,197 @@
+
+Assembly definition
+^^^^^^^^^^^^^^^^^^^
+
+Assembly defitinion should have the following form::
+
+    (.assembly Name
+        Item1
+        Item2
+        ...)
+
+Assembly items can be methods and types. Top level methods defined in an
+``.assembly`` form will be compiled to global CIL functions.
+
+Currently there is no support for type definition inside an assembly.
+
+Each assembly may contain one entry point method (either a static type method or
+an assembly global function marked by ``.entrypoint`` property).
+
+Method definition
+^^^^^^^^^^^^^^^^^
+
+Method definition should have the following form::
+
+    (.method Name (argument types) return-type (metadata items)
+        body-statements
+        ...)
+
+Method argument and return types should be fully-qualified (e.g. must include a
+namespace: for example, ``System.Void``).
+
+The only supported metadata item is ``.entrypoint``. It marks a method as an
+assembly entry point.
+
+Method example::
+
+    (.method Main () System.Void (.entrypoint)
+        (ldstr "Hello, world!")
+        (call (mscorlib System.Console WriteLine (System.String) System.Void))
+        (ret))
+
+Method body should be a CIL instruction sequence.
+
+CIL instructions
+^^^^^^^^^^^^^^^^
+
+Currently only a small subset of all available CIL instructions is supported by
+Naggum. This set will be extended in future.
+
+#. Call instruction::
+
+    (call (assembly type-name method-name (argument types) return-type))
+
+   Currently assembly name is ignored; only ``mscorlib`` methods can be called.
+   Static assembly function calls are not supported, but will be supported in
+   future.
+
+   Method argument and return types should be fully-qualified.
+
+#. Load string instruction::
+
+    (ldstr "Hello, world")
+
+   Loads a string to CLI stack.
+
+#. Return instruction::
+
+    (ret)
+
+   Return from current method.
+
+Example assembly definition
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+::
+
+    (.method Main () System.Void (.entrypoint)
+        (ldstr "Hello, world!")
+        (call (mscorlib System.Console WriteLine (System.String) System.Void))
+        (ret)))
+
+High level syntax
+-----------------
+
+Every high level Naggum program is a sequence of function definitions or a
+top-level executable statements. Functions may be called from top-level
+statements and will be available as a public static methods outside of an
+compiled assembly.
+
+Functions are defined using ``defun`` special form::
+
+    (defun function-name (arg1 arg2)
+        statement1
+        statement2)
+
+Currently executable statements may be one of the following.
+
+#. Arithmetic statements::
+
+    (+ 2 2)
+
+#. Function calls::
+
+    (defun func () (+ 2 2))
+
+    (func)
+
+#. Static CLI method calls::
+
+    (System.Console.WriteLine "Math:")
+
+#. If statements::
+
+    (if condition
+        true-statement
+        false-statement)
+
+#. Reduced if statements::
+
+    (if condition
```

Comment:
It is not named in CIL at all: there are multiple branching instructions, but none of them if called "if".

That's not the best place to discuss the language; this documentation just describes what we have. In future we will change and replace, I think, most of that, and will drop special meaning of Naggum.Compiler and high-level syntax completely.